torch/csrc/jit/fuser/cuda/fused_kernel.cpp

```diff
@@ -0,0 +1,187 @@
+#include "torch/csrc/jit/fuser/cuda/fused_kernel.h"
+
+#include "ATen/cuda/CUDAContext.h"
+#include "THC/THC.h"
+#include "torch/csrc/cuda/cuda_check.h"
+#include "torch/csrc/jit/resource_guard.h"
+
+// Note: unclear why this forward declaration is necessary
+#include "THC/THCTensorRandom.h"
+#include "THC/THCGenerator.hpp"
+THCGenerator* THCRandom_getGenerator(THCState* state);
+
+#include "nvrtc.h"
+#include "cuda.h"
+#include "cuda_runtime.h"
+
+#include <stdexcept>
+#include <sstream>
+#include <tuple>
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+namespace torch { namespace jit { namespace fuser { namespace cuda {
+
+void checkCUDAVersion(
+  const cudaDeviceProp& prop) {
+  if ((prop.major >= 6 && CUDA_VERSION < 8000) ||
+      (prop.major >= 7 && CUDA_VERSION < 9000)) {
+    std::stringstream err_string;
+    err_string << "In CUDAFusedKernel, PyTorch compiled with insufficient CUDA version: "
+         << CUDA_VERSION << " for the current GPU device " << prop.name
+         << " with device capability " << prop.major << "." << prop.minor;
+    throw std::runtime_error(err_string.str());
+  }
+}
+
+static void getMajorMinor(const cudaDeviceProp& prop, int& major, int& minor) {
+  int nvrtc_major, nvrtc_minor;
+  TORCH_NVRTC_CHECK(nvrtcVersion(&nvrtc_major, &nvrtc_minor));
+
+  // Short-circuits if NVRTC version too low
+  JIT_ASSERT(nvrtc_major >= 6);
+
+  // Major and minor is determined by device properties and 
+  // possibly "downcompiled" to a lower (compatible) compute architecture
+  // based on the NVRTC version
+  major = prop.major;
+  minor = prop.minor;
+  if (nvrtc_major <= 7 && prop.major > 5) { // 7 supports 2-5.x
+    major = 5;
+    if (prop.major == 5) minor = prop.minor;
```

Comment:
Nit: prop.major == 5 should never happen due to line 50.

Reply:
Same as above. Good catch.